### PR TITLE
Feat: Do initial training #4

### LINKED ISSUE #4: Initial Training
### Initial Training
We group this task in 5 subtasks:
- Data splitting
- Train Random Forest
- Optimise hyperparameters
- Best model selection
- Test

### Data Splitting
- [x] Use the component called train_val_test_split to extract and segregate the test set; its path is something like:
`mlflow.run(f"{config['main']['components_repository_src']}/train_val_test_split", ...)`
and add it to the pipeline of the main.py root file
- [x] There use the value configuration for the parameters like test_size, random_seed and stratify_by given by the modeling section of the config file; the MLproject file of the train_val_test_split directory delivers the parameter naming and description
- [x] Run the pipeline with
`> mlflow run . -P steps=data_split`
- [x] As a result in W&B the script is uploading 2 new datasets: _trainval_data.csv_ and _test_data.csv_

### Train Random Forest (RF)
- [x] Complete the script _src/train_random_forest/run.py_ according file instructions
- [x] Add the train_random_forest step to main.py. Use the name **random_forest_export** as output_artifact.
NOTE: the main.py file already provides a variable rf_config to be passed as the rf_config parameter.

### Optimise Hyperparameters
- [x] Re-run the entire pipeline varying the hyperparamters of the RF model by exploiting the Hydra configuration system.
Use the multi-run feature (adding the -m option at the end of the hydra_options specification).
Minimum config:  set modeling.max_tfidf_features to 10, 15 and 30, and modeling.random_forest.max_features to 0.1, 0.33, 0.5, 0.75, 1.
Remember: don't use spaces between hydra_options values! Or you will get the following information "... mismatched input '<EOF>' expecting {':', BRACKET_OPEN, BRACE_OPEN, FLOAT, INT, BOOL, NULL, UNQUOTED_CHAR, ID, ESC, WS, QUOTED_VALUE, INTERPOLATION}".

Example CLI command to get a coding impression:
```
> mlflow run . \
  -P steps=train_random_forest \
  -P hydra_options="modeling.random_forest.max_depth=10,50,100 modeling.random_forest.n_estimators=100,200,500 -m"
```
- [ ] Optional task: look at the Hydra documentation for even more ways to do hyperparameters optimisation

### Best Model Selection
- [x] In W&B select the best performing model
HINT: you should switch to the Table view (second icon on the left), then click on the upper right on "columns", remove all selected columns by clicking on "Hide all", then click on the left list on "ID", "Job Type", "max_depth", "n_estimators", "mae" and "r2". Click on "Close". Now in the table view you can click on the "mae" column on the three little dots, then select "Sort asc". This will sort the runs by ascending Mean Absolute Error (best result at the top).
When you have found the best job, click on its name, then go to its artifacts and select the "model_export" output artifact. You can now add a prod tag to it to mark it as "production ready".

### Test
Use the provided step **test_regression_mode**l to test your production model against the test set.
- [x] Implement the call to this component in the _main.py_ file with parameters given and described in the associated _MLproject_ file. Use the artifact _random_forest_export:prod_ for the parameter **mlflow_model** and the test artifact _test_data.csv:latest_ as **test_artifact**. 
Note: I created a new entry for the mlflow_model in the config.yaml: output_artifact with value "trained_random_forest_regression"
- [x] This test part does not run by default when you run the pipeline. It needs the manual step of promoting a model to prod before it can complete successfully. Activate it explicitly via CLI command:
`> mlflow run . -P steps=test_regression_model`

### PR DESCRIPTION
**Pull request by IloBe**

### CODE REVIEW

File: components/train_val_test_split/conda.yml
Comment (IloBe) [inline components/train_val_test_split/conda.yml:1]: newer wandb version necessary

File: components/train_val_test_split/run.py
Comment (IloBe) [inline components/train_val_test_split/run.py:1]: initial file

File: main.py
Comment (IloBe) [inline main.py:1]: added parts for initial training

File: src/train_random_forest/conda.yml
Comment (IloBe) [inline src/train_random_forest/conda.yml:1]: newer lib versions necessary

File: src/train_random_forest/feature_engineering.py
Comment (IloBe) [inline src/train_random_forest/feature_engineering.py:1]: changed doc parts

File: src/train_random_forest/run.py
Comment (IloBe) [inline src/train_random_forest/run.py:1]: initial file
